src/Development/IDE/Core/Tracing.hs

```diff
@@ -0,0 +1,166 @@
+getInstrumentCached :: IO (Maybe Key -> IO OurValueObserver)
+getInstrumentCached = do
+    instrumentMap <- newVar HMap.empty
+    mapBytesInstrument <- mkValueObserver "value map size_bytes"
+
+    let instrumentFor k = HMap.lookup k <$> readVar instrumentMap >>= \case
+            Nothing -> do
+                instrument <- mkValueObserver (BS.pack (show k ++ " size_bytes"))
+                modifyVar_ instrumentMap (return . HMap.insert k instrument)
+                return $ observe instrument
+            Just v -> return $ observe v
+    return $ maybe (return $ observe mapBytesInstrument) instrumentFor
+
+measureMemory :: [[Key]] -> (Maybe Key -> IO OurValueObserver) -> Var Values -> IO ()
+measureMemory groups instrumentFor stateRef = withSpan_ "Measure Memory" $ do
+    values <- readVar stateRef
+    valuesSizeRef <- newIORef 0
+    let !groupsOfGroupedValues = groupValues values
+    traceIO "STARTING MEMORY PROFILING"
+    forM_ groupsOfGroupedValues $ \groupedValues ->
+        repeatUntilJust $ do
+        traceIO (show $ map fst groupedValues)
+        runHeapsize $
+            forM_ groupedValues $ \(k,v) -> withSpan ("Measure " <> (BS.pack $ show k)) $ \sp -> do
+            acc <- liftIO $ newIORef 0
+            observe <- liftIO $ instrumentFor $ Just k
+            mapM_ (recursiveSize >=> \x -> liftIO (modifyIORef' acc (+ x))) v
+            size <- liftIO $ readIORef acc
+            let !byteSize = sizeOf (undefined :: Word) * size
+            setTag sp "size" (BS.pack (show byteSize ++ " bytes"))
+            () <- liftIO $ observe byteSize
+            liftIO $ modifyIORef' valuesSizeRef (+ byteSize)
+
+    valuesSize <- readIORef valuesSizeRef
+    observe <- instrumentFor Nothing
+    observe valuesSize
+    traceIO "MEMORY PROFILING COMPLETED"
+    where
+        blacklist = []
+
+        groupValues :: Values -> [ [(Key, [Value Dynamic])] ]
+        groupValues values =
+            let groupedValues =
+                    [ [ (k, vv)
+                      | k <- groupKeys
+                      , let vv = [ v | ((_,k'), v) <- HMap.toList values , k == k']
+                      ]
+                    | groupKeys <- groups
+                    ]
+                otherValues =
+                    HMap.toList $
+                    HMap.fromListWith (++)
+                    [ (k, [v])
+                    | ((_, k), v) <- HMap.toList values
+                    , k `notElem` blacklist
+                    , k `notElem` concat groups
+                    ]
+                !result = groupedValues <> map (:[]) otherValues
+                -- force the spine of the nested lists
+            in result `using` seqList (seqList (seqTuple2 r0 (seqList r0)))
+
+groupForObservableSharing :: [[Key]]
+groupForObservableSharing =
+        -- Group keys in bundles to be analysed jointly (for sharing)
+        -- Ideally all the keys would be analysed together, but if we try to do too much the GC will interrupt the analysis
```

Comment:
I think GC can't run while there is a `unsafe` FFI call is in progress, so maybe we can fork a thread that calls a C function which waits for the analysis to complete (as signaled by a write to the file system or something).

Reply:
This is correct, but the Haskell threads do not run either. 

The only way imho to take advantage of this property is to rewrite the heap size analysis in C

Reply:
Andreas tells me on #ghc that this is not true, and Haskell threads will continue to run as long we are using the threaded runtime.

Reply:
I think that's probably a misunderstanding between you and Andreas, but I'm happy to be corrected by a program that demonstrates your theory. 

In the meantime, I'll quote https://downloads.haskell.org/ghc/latest/docs/html/users_guide/runtime_control.html#rts-flag---long-gc-sync=<seconds>